### PR TITLE
Termination option : requested changes

### PR DESCRIPTION
**Pull request by GuillaumeGen**
Added some docs and factor out part of the code.

### CODE REVIEW

File: doc/sections/options.md [hunk 1]
```diff
@@ -51,7 +51,7 @@ When in default mode, the following flags are available for configuration:
 
 #### Confluence checking
 
-Confluence checking (and also termination) must be established for each of the
+Confluence (and also termination) must be established for each of the
 considered rewriting systems contained in `lambdapi` files. By default,  these
```
Comment (fblanqui) [inline doc/sections/options.md]: each ... system-s
must -> Lambdapi provides an option to check the confluence of the set of rewriting rules declared in a file.

Question: what is this set? Is it the smallest set obtained by adding all the rules in the file and its dependencies? What if several file names are given in argument?

File: doc/sections/options.md [hunk 2]
```diff
@@ -78,7 +78,33 @@ lambdapi --confluence "cat > output.trs; echo MAYBE" input_file.lp
 
 #### Termination checking
 
-For now, there is no support for termination checking.
+Like confluence, termination must be established for each of the
```
Comment (fblanqui) [inline doc/sections/options.md]: Start like for confluence.

File: src/confluence.ml [hunk 3]
```diff
@@ -10,7 +10,6 @@ open Terms
 
 (** Logging function for the confluence checker interface. *)
 let log_conf = new_logger 'c' "conf" "informations for the confluence checker"
-let log_conf = log_conf.logger
```
Comment (fblanqui) [inline src/confluence.ml]: Why do you remove this line?

File: src/external_checker.ml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+(** [check cmd sign translator logging property] runs the checker specified
+    by command [cmd] on the rewrite system of signature [sign] tranlated by
+    [translator]. The return value is [Some true] in the case where the system
+    is accepted by [cmd]. It is [Some false] if the system is refused by
+    [cmd], and it is [None] if the tool cannot conclude.  Note that it is
+    assumed that [cmd] corresponds to a command that outputs either ["YES"],
+    ["NO"] or ["MAYBE"] as the first line of its standard output. The
+    exception [Fatal] may be raised if [cmd] does not behave as expected.
+    [checked_property] is for printing purpose *)
+let check : string -> Sign.t -> (Format.formatter -> Sign.t -> unit) ->
```
Comment (fblanqui) [inline src/external_checker.ml]: Give a name to the type Format.formatter -> 'a -> unit ?

**Comment by fblanqui** [comment]
Thanks Guillaume. It looks nice to me. You just need to fix a few things. See my comments.

File: doc/sections/options.md (hunk 1 above)
Comment (GuillaumeGen) [reply doc/sections/options.md]: The set of rewrite rules considered is the set of all rules declared in any of the file of the transitive closure computed via the `require`.
I don't understand what you mean by : 
> must -> Lambdapi provides an option to check the confluence of the set of rewriting rules declared in a file by calling external provers.
Your suggestion seems to be a reformulation of the next sentence. So does it mean "suppress the sentence containing must and replace the next one by"?

**Comment by fblanqui** [comment]
Yes, I suggest to start by "Lambdapi provides an option to check... " 
and don't use "must" because it's optional.

File: src/termination.ml [hunk 5]
```diff
@@ -136,8 +135,8 @@ let print_tl_rule : Format.formatter -> int -> sym -> rule -> unit =
   Format.fprintf oc "<TLrhs>@.%a</TLrhs>@.</typeLevelRule>@."
     (print_type i s.sym_name) rhs
 
-(** [get_vars s r] returns the list of variable of used in the rule [r],
-    in the form of a couple containing the name of the variable and its type,
+(** [get_vars s r] returns the list of variables used in the rule [r],
+    in the form of a pair containing the name of the variable and its type,
     inferred by the solver. *)
 let get_vars : sym -> rule -> (string * Terms.term) list = fun s r ->
   (*check_rule s r;*)
```
Comment (fblanqui) [inline src/termination.ml]: What is this comment for?

File: doc/sections/options.md [hunk 6]
```diff
@@ -78,7 +76,31 @@ lambdapi --confluence "cat > output.trs; echo MAYBE" input_file.lp
...
+rules declared in a file by calling external provers using the XTC format.
+The `--termination <cmd>` flag specifies the termination-checking command to
+be used. The command is expected to behave as follows:
+ - take the problem description (in `.xml` format) on its standard input,
+ - output on its first line either `YES`, `NO` or `MAYBE`.
+
+As an example,  `echo MAYBE` is the simplest possible (valid)
+termination-check that one may use.
+
+As far as we now, `SizeChangeTool` is the only termination checker compatible
```
Comment (fblanqui) [inline doc/sections/options.md]: Add URL.

File: doc/sections/options.md [hunk 7]
```diff
@@ -51,11 +51,9 @@ When in default mode, the following flags are available for configuration:
 
 #### Confluence checking
 
-Confluence checking (and also termination) must be established for each of the
-considered rewriting systems contained in `lambdapi` files. By default,  these
-checks are not performed, and they must be explicitly requested.
 
-We provide an interface to external confluence checkers using the TRS  format.
+Lambdapi provides an option to check the confluence of the set of rewriting
+rules declared in a file by calling external provers using the TRS  format.
```
Comment (fblanqui) [inline doc/sections/options.md]: Add URL for CSI^ho.

File: doc/sections/options.md [hunk 8]
```diff
@@ -78,7 +76,31 @@ lambdapi --confluence "cat > output.trs; echo MAYBE" input_file.lp
...
+termination-check that one may use.
+
+As far as we now, `SizeChangeTool` is the only termination checker compatible
+with all `lambdapi` features.
+It can be called in the following way.
+```bash
+lambdapi --termination "path/to/sct.native -nc --stdin=xml" input_file.lp
+```
+
+If the file does not contain type-level rewriting, `Wanda` is also compatible.
```
Comment (fblanqui) [inline doc/sections/options.md]: Add URL to Wanda.

File: doc/sections/options.md (hunk 7 above)
Comment (fblanqui) [inline doc/sections/options.md]: URL to TRS format missing.

File: doc/sections/options.md [hunk 9]
```diff
@@ -78,7 +78,33 @@ lambdapi --confluence "cat > output.trs; echo MAYBE" input_file.lp
 
 #### Termination checking
 
-For now, there is no support for termination checking.
+Lambdapi provides an option to check the termination of the set of rewriting
+rules declared in a file by calling external provers using the XTC format.
```
Comment (fblanqui) [inline doc/sections/options.md]: URL to XTC format missing.

File: src/termination.ml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 (** Temination checking.
 
-    This module allows the translation of a signature in the TPDB format (used
+    This module allows the translation of a signature in the XTC format (used
     in the termination competition). See the "Termination Problem Database":
     http://cl2-informatik.uibk.ac.at/mercurial.cgi/TPDB *)
```
Comment (fblanqui) [inline src/termination.ml]: This is not exactly the URL of the XTC format.

Comment (GuillaumeGen) [reply src/termination.ml]: But people reading this line probably want to see the Problem Data Base rather than an obscure XSD file.

**Pull request merged**